pwndbg/dbg/gdb.py

```diff
@@ -17,6 +17,61 @@
 from pwndbg.gdblib import load_gdblib
 
 
+class GDBFrame(pwndbg.dbg_mod.Frame):
+    def __init__(self, inner: gdb.Frame):
+        self.inner = inner
+
+    @override
+    def evaluate_expression(self, expression: str) -> pwndbg.dbg_mod.Value:
+        selected = gdb.selected_frame()
+        restore = False
+        if selected != self.inner:
+            self.inner.select()
+            restore = True
+
+        try:
+            value = gdb.parse_and_eval(expression, False)
+        except TypeError:
+            # Some earlier versions of GDB don't support the global_context argument.
+            value = gdb.parse_and_eval(expression)
+        finally:
+            if restore:
+                selected.select()
+
+        return GDBValue(value)
+
+
+class GDBThread(pwndbg.dbg_mod.Thread):
+    def __init__(self, inner: gdb.InferiorThread):
+        self.inner = inner
+
+    @override
+    def bottom_frame(self) -> pwndbg.dbg_mod.Frame:
+        selected = gdb.selected_thread()
+        restore = False
+        if selected != self.inner:
+            self.inner.switch()
+            restore = True
+
+        value = gdb.newest_frame()
+        if restore:
+            selected.switch()
+        return GDBFrame(value)
+
+
+class GDBProcess(pwndbg.dbg_mod.Process):
+    def __init__(self, inner: gdb.Inferior):
+        self.inner = inner
+
+    @override
+    def evaluate_expression(self, expression: str) -> pwndbg.dbg_mod.Value:
+        try:
+            return GDBValue(gdb.parse_and_eval(expression, True))
+        except TypeError:
+            # Some earlier versions of GDB don't support the global_context argument.
+            return GDBValue(gdb.parse_and_eval(expression))
```

Comment:
Since we do this twice in this file, it might be good to have a helper method in this function that does this for us, and use it here and in `GDBFrame`?

Reply:
Yeah, that makes sense. I'll put them both in a single helper function.